newton/sim/builder.py

```diff
@@ -3507,6 +3516,73 @@ def finalize(self, device: Devicelike | None = None, requires_grad: bool = False
+                            self.body_mass[i] = corrected_mass
+                            self.body_inertia[i] = corrected_inertia
+                            # Update inverse mass and inertia
+                            if corrected_mass > 0.0:
+                                self.body_inv_mass[i] = 1.0 / corrected_mass
+                            else:
+                                self.body_inv_mass[i] = 0.0
+
+                            if any(x for x in corrected_inertia):
+                                self.body_inv_inertia[i] = wp.inverse(corrected_inertia)
+                            else:
+                                self.body_inv_inertia[i] = corrected_inertia
+                else:
+                    # Use fast Warp kernel validation
+                    # First create arrays for the kernel
+                    body_mass_array = wp.array(self.body_mass, dtype=wp.float32, requires_grad=False)
+                    body_inertia_array = wp.array(self.body_inertia, dtype=wp.mat33, requires_grad=False)
+                    body_inv_mass_array = wp.array(self.body_inv_mass, dtype=wp.float32, requires_grad=False)
+                    body_inv_inertia_array = wp.array(self.body_inv_inertia, dtype=wp.mat33, requires_grad=False)
+                    correction_flags = wp.zeros(len(self.body_mass), dtype=wp.int32)
+
+                    # Launch validation kernel
+                    wp.launch(
+                        kernel=validate_and_correct_inertia_kernel,
+                        dim=len(self.body_mass),
+                        inputs=[
+                            body_mass_array,
+                            body_inertia_array,
+                            body_inv_mass_array,
+                            body_inv_inertia_array,
+                            self.balance_inertia,
+                            self.bound_mass if self.bound_mass is not None else 0.0,
+                            self.bound_inertia if self.bound_inertia is not None else 0.0,
+                            correction_flags,
+                        ],
+                    )
+
+                    # Check if any corrections were made
+                    num_corrections = int(np.sum(correction_flags.numpy()))
+                    if num_corrections > 0:
+                        warnings.warn(
+                            f"Inertia validation corrected {num_corrections} bodies. "
+                            f"Set validate_inertia_detailed=True for detailed per-body warnings.",
+                            stacklevel=2,
+                        )
+
+                    # Copy corrected values back
+                    self.body_mass = body_mass_array.numpy().tolist()
+                    self.body_inertia = [wp.mat33(m) for m in body_inertia_array.numpy()]
+                    self.body_inv_mass = body_inv_mass_array.numpy().tolist()
+                    self.body_inv_inertia = [wp.mat33(m) for m in body_inv_inertia_array.numpy()]
```

Comment:
I don't think we need those updated values in the ModelBuilder here, I think we expect `ModelBuilder.finalize()` to give us a Model and not necessarily a corrected ModelBuilder. The `validate_inertia_detailed` pass makes sense to update the builder entries but for the fast pass we can skip this step. It would be good to include this in the docstrings though.